cmd/protoc-gen-crd/openapiGenerator.go

```diff
@@ -0,0 +1,829 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this currentFile except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"log"
+	"slices"
+	"strings"
+
+	"github.com/getkin/kin-openapi/openapi3"
+	"github.com/golang/protobuf/protoc-gen-go/descriptor"
+	plugin "github.com/golang/protobuf/protoc-gen-go/plugin"
+	"golang.org/x/exp/maps"
+	"google.golang.org/protobuf/proto"
+	apiextinternal "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	apiext "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	crdmarkers "sigs.k8s.io/controller-tools/pkg/crd/markers"
+	"sigs.k8s.io/controller-tools/pkg/markers"
+	"sigs.k8s.io/yaml"
+
+	"istio.io/tools/cmd/protoc-gen-crd/pkg/protomodel"
+)
+
+// Some special types with predefined schemas.
+// This is to catch cases where solo apis contain recursive definitions
+// Normally these would result in stack-overflow errors when generating the open api schema
+// The imperfect solution, is to just generate an empty object for these types
```

Comment:
Consistency with other remarks:
```suggestion
// Some special types with predefined schemas.
// This is to catch cases where @solo.io apis contain recursive definitions.
// Normally these would result in stack-overflow errors when generating the open api schema.
// The imperfect solution, is to just generate an empty object for these types.
```